File: backend/app/api/v1/customers.py
import uuid

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_user, require_admin
from app.core.database import get_db
from app.models.user import User
from app.schemas.customer import CustomerCreate, CustomerListOut, CustomerOut, CustomerUpdate
from app.services import customer_service

router = APIRouter(prefix="/customers", tags=["customers"])


@router.post("", response_model=CustomerOut, status_code=201)
async def create_customer(
    data: CustomerCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return await customer_service.create_customer(db, data, created_by=current_user)


@router.get("", response_model=list[CustomerListOut])
async def list_customers(
    search: str | None = None,
    blocked_only: bool = False,
    include_inactive: bool = False,
    db: AsyncSession = Depends(get_db),
    _: User = Depends(get_current_user),
):
    return await customer_service.list_customers(
        db,
        active_only=not include_inactive,
        search=search,
        blocked_only=blocked_only,
    )


@router.get("/{customer_id}", response_model=CustomerOut)
async def get_customer(
    customer_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    _: User = Depends(get_current_user),
):
    customer = await customer_service.get_customer(db, customer_id)
    if not customer:
        raise HTTPException(status_code=404, detail="Cliente não encontrado")
    return customer


@router.patch("/{customer_id}", response_model=CustomerOut)
async def update_customer(
    customer_id: uuid.UUID,
    data: CustomerUpdate,
    db: AsyncSession = Depends(get_db),
    _: User = Depends(get_current_user),
):
    customer = await customer_service.get_customer(db, customer_id)
    if not customer:
        raise HTTPException(status_code=404, detail="Cliente não encontrado")
    return await customer_service.update_customer(db, customer, data)

File: frontend/src/api/customers.ts
import { api } from './client'
import type { Customer } from '@/types'

export interface CustomerCreate {
  name: string
  phone?: string
  document?: string
  address?: string
  customer_type: string
  credit_limit?: number
  notes?: string
}

export interface CustomerListParams {
  search?: string
  blocked_only?: boolean
  include_inactive?: boolean
}

export async function listCustomers(params?: CustomerListParams): Promise<Customer[]> {
  const { data } = await api.get<Customer[]>('/customers', { params })
  return data
}

export async function getCustomer(id: string): Promise<Customer> {
  const { data } = await api.get<Customer>(`/customers/${id}`)
  return data
}

export async function createCustomer(body: CustomerCreate): Promise<Customer> {
  const { data } = await api.post<Customer>('/customers', body)
  return data
}

export async function updateCustomer(id: string, body: Partial<CustomerCreate> & { is_active?: boolean }): Promise<Customer> {
  const { data } = await api.patch<Customer>(`/customers/${id}`, body)
  return data
}
